### ISSUE TITLE
Challenge 20 - DSS MLCosting for requests

### ISSUE THREAD

**Issue by DiegoDomenici** [ECMWFCode4Earth/Challenges_2025#12]
# Challenge 20 - DSS MLCosting for requests

> **Stream 2 - Machine Learning for Earth Sciences Applications**

### Goal
The objective of this work is to develop an MLCosting based plug-in component that can be seamlessly integrated with the existing DSS  requests costing function as an API.  This function will receive incoming requests as input, estimate cost based on a ML algorithm trained with historical track of requests and return a value (numerical of a range of categories) to be used by the QoS to evaluate priorities.

### Mentors
Angel Lopez, Gionata Biavati, Petrut Cobarzan (all ECMWF)

### Skills Required
* Python
* Statistical Packages/ML Algorithms

<hr>

### Description

The objective of this work is to develop an ML based plug-in component that can be seamlessly integrated with the existing costing function as an API. This plug-in will operate as an independent, non-essential module, allowing it to be activated or deactivated as needed as per adaptor and catalogued dataset.  

This costing API will leverage Machine Learning to analyze incoming requests and provide inference-based classifications to assist our broker in determining the optimal number of processes to execute for a given payload. 

ML models are expected to be feed and trained with multiple information sources such as the structure and configuration of raw data sources and the historical track of requests.  

<hr>

### Evaluation Criteria
* Feasibility
* Innovative approach
* Transferability

**Comment by thinhngo-x** [comment]
Dear mentors and organizers of the challenge,

I am very interested in your project and would like to clarify a few points to better understand the scope and requirements:

1. Could you please provide more details on the available data, including its format? A small piece of sample data could definitely help.
2. Since you mentioned "provide inference-based classifications", I suspect that this could be a classification problem. Therefore, what are the particular classes that you have in mind and how could they help optimize resource allocation?
3. Do you have in mind any existing work (research articles or POCs) that is close enough to this project to which I can refer?

Thank you for your time. I look forward to your response.

**Comment by ALopAlos** [comment]
Hi, thanks a lot for your interest. Really looking forward for your proposal.
1) Currently data about requests can be provided via different formats. For the requests we can give you a CSV (with some fields which are JSONs) extracted or from the transactional data base or from the logs that we ingest in Splunk.  Basically this contains information about the requests structure, timing duration per step, scope, user details, ... Examples of files will be provided.
2) Our system relies on a Quality of Service (QoS) component that allow as via simple rules ( priority, user, limits) to manage the traffic of the system by different criteria. Our aim is to get the envisaged costing of a requests based on history for same type of request, type of user, time of the day, status of the queue, available slots on data source, ... In principle the idea is the Algorithm to receive the incoming requests and based on the previous knowledge of the model be tagged with a standard category ( light - normal - heavy - denied ). This category will be used by QoS to manage the request in the queue and when to run it. In principle we could constraint the scope of the pilot to the most demanded datasets. You can have an estimate on https://cds.climate.copernicus.eu/live or https://ads-test-cci2.copernicus-atmosphere.eu/live. 
3) Not at hand. We will check and let you know.

**Comment by ecmwf-cobarzan** [comment]
Hi @thinhngo-x.

To get a feel on how requests look like, you can go to any of the datasets download forms under https://cds.climate.copernicus.eu/datasets (e.g. https://cds.climate.copernicus.eu/datasets/reanalysis-era5-pressure-levels?tab=download), do a random selection and then go at the bottom of the page and press "Show API request code". The requests that we process consist of the pairs of dataset name + request dictionary, as in the generated code snippet (perhaps presented to you in a JSON format). When you formulate a valid request the "Submit form" button at the bottom of the page becomes active. If you execute the request (from the webportal, by pressing the button), you'll eventually see the result in the "Your requests" page. Assuming that the request was successful, you'll be able to see how long the whole request took to execute and how large the result file is.

The input to be fed into the neural network (or other ML mechanism) will consist of lists of tuples containing:
- the dataset name
- the request dictionary
- the execution time
- the size of the result
- possibly other metadata.

Additionally, we already have an implementation of the costing that you can see on the right column in the download form. As you do your selection (if your selection is large enough) you'll see the cost estimate on the "Request size" bar (as a percentage of the maximum allowed cost). The classes described by Angel will be more or less some intervals on this bar (and they'll be dataset specific).

Let us know if you need any other details.
Thanks for your interest!

**Comment by thinhngo-x** [comment]
Thank you @ALopAlos and @ecmwf-cobarzan for your detailed replies.

As far as I can understand, the issue is that when one user request for a dataset with some specific parameters, the server will process the request and hand out the requested data for the user to download. I'm still wondering whether the payload execution you mentioned is to process the data request or to handle the downloading process?

I'm also a little bit confused about the input data that @ecmwf-cobarzan mentioned. Would the execution time and the size of the result be in input as well? Or you mean they are labels for training data? If they are in input data, it means that there is a system that can estimate execution time and dataset size given the request dictionary? Otherwise, during inference time, we have to wait for the data processing to finish to retrieve these features and input them in the ML model.

Finally, talking about request dictionary, as far as I see, it has a "loosen" schema right? Since the set of request parameters vary per dataset?

Thank you again for your time and patience,

**Comment by ecmwf-cobarzan** [comment]
Hi @thinhngo-x.

> I'm still wondering whether the payload execution you mentioned is to process the data request or to handle the downloading process?

It is to process the data request (which behind the scenes involved pre-processing steps, data retrieval from the data provider and post-processing steps). The effort for the download has not been yet factored into out costing estimates, but this could be considered. However, this is likely easy to estimate as it is more or less proportional to the size of the result file(s). In comparison, the complexity of the request processing makes it significantly harder to estimate its costing.

> Would the execution time and the size of the result be in input as well? Or you mean they are labels for training data?

They will be labels for the training (i.e. the expected output for the corresponding requests). We have plenty of such pairs from our production environment, we do not need to wait for new requests to produce new ones.

> Finally, talking about request dictionary, as far as I see, it has a "loosen" scheme right? Since the set of request parameters vary per dataset?

I am not sure what you mean, but every dataset has its own set of parameters/dimensions. Some datasets do resemble, but this is exercise it meant to produce dataset specific estimates.

Let us know if you need any other details.
Thanks for your interest!

**Comment by thinhngo-x** [comment]
> I am not sure what you mean, but every dataset has its own set of parameters/dimensions. Some datasets do resemble, but this is exercise it meant to produce dataset specific estimates. I recommend you play a bit with the download forms in the data stores, to see some actual datasets.

I meant that every dataset has its own set of parameters. Anyway, you have confirmed it already.
We are working on the proposal right now. Thanks again for your response.
